private/couchbase/transactions.hxx

```diff
@@ -0,0 +1,258 @@
+ *
+ * For a much more detailed example, see @ref examples/game_server.cxx
+ *
+ * @section txn_best_practices Best Practices
+ *
+ * Each @ref transactions instances spins up background threads to perform cleanup of metadata
+ * that could be left behind after failed transactions.  For that reason, creating many
+ * @ref transactions objects, especially if they are long-lived, will consume resources.  We recommend
+ * simply creating one @ref transactions object per process, and using that for the life of the process,
+ * when possible.
+ *
+ * The @ref transactions, @ref cluster, @ref bucket, and @ref collection instances are all safe to use
+ * across threads.  The @ref cluster and @ref bucket use libcouchbase internally, and manage a pool of
+ * libcouchbase instances, since those instances are not to be used across threads.  The maximum number
+ * of instances in the pools (for either the @ref cluster or @ref bucket) can be specified at the time
+ * you construct the @ref cluster.  See @ref cluster_options for details.
+ *
+ * Since a @ref transactions instance spins up one thread per bucket for background cleanup tasks, a reasonable
+ * starting point for sizing the @ref cluster_options::max_bucket_instances would be one per bucket, plus
+ * one per thread that uses the @ref cluster (or the @ref transactions object constructed with that @ref cluster).
+ * This could be tuned lower potentially, but this is a the maxmimum one should need.  The pool only
+ * creates a new instance when it needs one, so the actual number created may never hit the max.  When
+ * the maximum is reached, the thread will block until one is available.
+ *
+ * @example examples/game_server.cxx
+ */
+
+/** @brief Main class for creating a transaction
+ *
+ */
+class transactions
+{
+  public:
+    /**
+     * @brief Create a transactions object.
+     *
+     * Creates a transactions object, which can be used to run transactions within the current thread.
+     *
+     * @param cluster The cluster to use for the transactions.
+     * @param config The configuration parameters to use for the transactions.
+     */
+    transactions(std::shared_ptr<core::cluster> cluster, const transaction_config& config);
+
+    /**
+     * @brief Destructor
+     */
+    ~transactions();
+
+    /**
+     * @brief Run a transaction
+     *
+     * Expects a lambda, which it calls with an @ref attempt_context reference to be used in the lambda for
+     * the transaction operations.
+     *
+     * @param logic The lambda containing the transaction logic.
+     * @return A struct containing some internal state information about the transaction.
+     * @throws @ref transaction_failed, @ref transaction_expired, @ref transaction_commit_ambiguous, all of which
+     *         share a common base class @ref transaction_exception.
+     */
+    transaction_result run(logic&& code);
+
+    transaction_result run(const per_transaction_config& config, logic&& code);
```

Comment:
In cxx, it is more readable to have the lambda at the end.   However, this could be flipped around, with a default config...   I added a ticket for this CXXCBC-228